rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1614 @@
+then
+  { inherit path matches; }
+else if path == /. then
+  null
+else
+  go (dirOf path);
+```
+
+**Alternatives**
+
+- The bodies could be absorbed in some cases, saving an indentation level:
+  ```nix  
+  #1a
+  if builtins.length matches != 0 then {
+    inherit path matches;
+  } else if path == /. then [
+    1
+    2
+  ] else
+    go (dirOf path);
+  ```
+  - This results in inconsistent vertical start of the keywords, making the structure harder to follow
+- Have the `then` on the start of the next line, directly followed by the if body:
+  ```nix
+  #1b
+  if builtins.length matches != 0
+    then { inherit path matches; }
+  else if path == /.
+    then [
+      1
+      2
+    ]
+  else go (dirOf path);
+
+  #1c
+  if builtins.length matches != 0
+  then { inherit path matches; }
+  else if path == /.
+  then [
+    1
+    2
+  ]
+  else go (dirOf path);
+  ```
+
+### assert
+
+- `assert <cond>;` mirrors the formatting for [bindings](#bindings).
+- `assert`s `<body>` must always start on their own line and the body also starts on its own line without any additional indentation.
+
+```nix
+# Good
+assert foo;
+[
+  bar
+  baz
+]
+
+# Bad
+assert foo; [
+  bar
+  baz
+]
+
+# Good
+{
+  vendor ?
+    assert false;
+    null,
+    
+  vendor ? null,
+}:
+null
+
+let
+  # Good
+  x =
+    assert foo;
+    bar;
+   
+  # Bad
+  y = assert foo;
+    bar;
+in
+x
+
+# Multiline condition
+assert
+  let
+    x = true;
+  in x;
+  true
+
+# Function call condition with absorbed last argument, same formatting as bindings
+assert assertMsg (isPath path) ''
+  lib.path.append:
+    The first argument is of type ${builtins.typeOf path}, but a path was expected
+'';
+true
+```
+
+**Alternatives**
+- Treat it the [same as `with`](#with). The reasons not to do that:
+  - `assert`'s stand on their own and could be removed without breaking anything. Comparatively, `with`'s can't be removed without breaking the code
+  - `assert`'s are a bit like `if-then-else` statements, which are also spread out over multiple lines
+
+### with
+
+- In any situation where a term would get absorbed, the term with a `with` prepended to it may get absorbed as well.
+- Otherwise, the body of `with attrs;` must start on a new line without any additional indentation.
+
+**Examples**
+
+```nix
+{
+  # Good
+  foo = with bar; [
+    # multiline
+    baz
+  ];
+  
+  # Good
+  foo =
+    with foo;
+    with bar;
+    [
+      # multiline
+      baz
+    ];
+  
+  # Good
+  foo =
+    with bar;
+    baz foo {
+      # multiline
+      qux = 10;
+    };
+
+  # Good
+  foo =
+    with bar;
+    if cond then
+      foo
+    else
+      bar;
+  
+  # Bad
+  foo = assert qux; with bar; [
+    # multiline
+    baz
+  ];
+  
+  # Bad
+  foo = with bar;
+    [
+      # multiline
+      baz
+    ];
+
+  # Bad
+  foo =
+    with bar; [
+      # multiline
+      baz
+    ];
+    
+  # Good
+  [
+    qux
+    quux
+  ]
+  ++ (with pkgs; [
+    baz
+    blorp
+  ]);
+}
+```
+
+### let-in
```

Comment:
I don't know if this was discussed already, but what about this example?

```
# let in the same line

foo = let
  A = "oof";
in
  A+A;

# let in the next line -- one more indentation
foo =
  let
    A = "oof";
  in
    A+A;
```

Is this expected?

Reply:
See the bindings absorption rules for reference.

> Within bindings, if the first and last line are not indented, the absorbed style is used, otherwise newline and indent

The last line is always indented in a let binding.